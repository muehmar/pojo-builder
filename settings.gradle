rootProject.name = "pojo-builder"

include 'core'
include 'core-annotations'
include 'example'
include 'example-java17'

dependencyResolutionManagement {
    versionCatalogs {
        libs {
            version('pojobuilder', '1.6.0')
            library('pojobuilder-core', 'io.github.muehmar','pojo-builder').versionRef('pojobuilder')
            library('pojobuilder-annotations', 'io.github.muehmar','pojo-builder-annotations').versionRef('pojobuilder')

            library('codeGenerator', 'io.github.muehmar:code-generator:0.22.0')

            version('junit', '5.10.0')
            library('junit-api', 'org.junit.jupiter', 'junit-jupiter-api').versionRef('junit')
            library('junit-engine', 'org.junit.jupiter', 'junit-jupiter-engine').versionRef('junit')
            library('junit-params', 'org.junit.jupiter', 'junit-jupiter-params').versionRef('junit')

            version('assertj', '3.26.3')
            library('assertj', 'org.assertj', 'assertj-core').versionRef('assertj')

            bundle('junit', ['junit-api', 'junit-engine', 'junit-params', 'assertj'])

            library('jakarta-annotationApi', 'jakarta.annotation:jakarta.annotation-api:2.1.1')

            library('google-autoservice', 'com.google.auto.service:auto-service:1.1.1')
            library('joor', 'org.jooq:joor-java-8:0.9.15')

            library('lombok', 'org.projectlombok:lombok:1.18.32')

            library('snapshotTesting','io.github.origin-energy:java-snapshot-testing-junit5:4.0.8')

            library('slf4j', 'org.slf4j:slf4j-api:2.0.7')

            plugin('release', 'net.researchgate.release').version('2.8.1')
            plugin('versions', 'com.github.ben-manes.versions').version('0.51.0')
        }
    }
}