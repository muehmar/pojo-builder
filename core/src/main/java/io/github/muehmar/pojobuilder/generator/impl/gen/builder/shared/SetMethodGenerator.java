package io.github.muehmar.pojobuilder.generator.impl.gen.builder.shared;

import static io.github.muehmar.codegenerator.java.JavaModifier.PUBLIC;
import static io.github.muehmar.codegenerator.java.MethodGen.Argument.argument;
import static io.github.muehmar.pojobuilder.generator.impl.gen.Generators.newLine;
import static io.github.muehmar.pojobuilder.generator.impl.gen.Refs.JAVA_UTIL_OPTIONAL;

import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.java.JavaGenerators;
import io.github.muehmar.codegenerator.java.JavaModifier;
import io.github.muehmar.codegenerator.java.JavaModifiers;
import io.github.muehmar.pojobuilder.generator.impl.gen.RefsGen;
import io.github.muehmar.pojobuilder.generator.model.Pojo;
import io.github.muehmar.pojobuilder.generator.model.PojoAndField;
import io.github.muehmar.pojobuilder.generator.model.settings.PojoSettings;
import java.util.function.BiFunction;

public class SetMethodGenerator {
  private SetMethodGenerator() {}

  public static Generator<Pojo, PojoSettings> setMethodGenerator(
      BiFunction<Pojo, PojoSettings, String> builderClassName, ModifierOption modifierOption) {
    return Generator.<Pojo, PojoSettings>emptyGen()
        .appendList(
            standardSetMethod(builderClassName, modifierOption), Pojo::getPojoAndFields, newLine())
        .appendSingleBlankLine()
        .appendList(
            optionalSetMethod(builderClassName).appendSingleBlankLine(), Pojo::getPojoAndFields);
  }

  private static Generator<PojoAndField, PojoSettings> standardSetMethod(
      BiFunction<Pojo, PojoSettings, String> builderClassName, ModifierOption modifierOption) {
    final Generator<PojoAndField, PojoSettings> content =
        (paf, settings, writer) ->
            writer
                .println("this.%s = %s;", paf.getField().getName(), paf.getField().getName())
                .println("return this;");

    return JavaGenerators.<PojoAndField, PojoSettings>methodGen()
        .modifiers(
            (paf, s) ->
                JavaModifiers.of(
                    paf.getField().isRequired()
                            && modifierOption.equals(ModifierOption.REQUIRED_PRIVATE)
                        ? JavaModifier.PRIVATE
                        : JavaModifier.PUBLIC))
        .noGenericTypes()
        .returnType(
            (paf, s, w) ->
                w.println(
                    builderClassName.apply(paf.getPojo(), s)
                        + paf.getPojo().getTypeVariablesFormatted()))
        .methodName((paf, s) -> paf.getField().builderSetMethodName(s).asString())
        .singleArgument(
            paf ->
                argument(paf.getField().getType().getTypeDeclaration(), paf.getField().getName()))
        .doesNotThrow()
        .content(content)
        .build()
        .append(RefsGen.fieldRefs(), PojoAndField::getField);
  }

  private static Generator<PojoAndField, PojoSettings> optionalSetMethod(
      BiFunction<Pojo, PojoSettings, String> builderClassName) {
    final Generator<PojoAndField, PojoSettings> content =
        (paf, settings, writer) ->
            writer
                .println(
                    "this.%s = %s.orElse(null);",
                    paf.getField().getName(), paf.getField().getName())
                .println("return this;");

    return JavaGenerators.<PojoAndField, PojoSettings>methodGen()
        .modifiers(PUBLIC)
        .noGenericTypes()
        .returnType(
            (paf, s, w) ->
                w.print(
                    builderClassName.apply(paf.getPojo(), s)
                        + paf.getPojo().getTypeVariablesFormatted()))
        .methodName((paf, s) -> paf.getField().builderSetMethodName(s).asString())
        .singleArgument(
            paf ->
                argument(
                    String.format("Optional<%s>", paf.getField().getType().getTypeDeclaration()),
                    paf.getField().getName()))
        .doesNotThrow()
        .content(content)
        .build()
        .append(w -> w.ref(JAVA_UTIL_OPTIONAL))
        .append(RefsGen.fieldRefs(), PojoAndField::getField)
        .filter(pf -> pf.getField().isOptional());
  }

  public enum ModifierOption {
    PUBLIC,
    REQUIRED_PRIVATE
  }
}
