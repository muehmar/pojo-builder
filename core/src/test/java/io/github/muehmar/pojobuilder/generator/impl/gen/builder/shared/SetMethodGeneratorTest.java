package io.github.muehmar.pojobuilder.generator.impl.gen.builder.shared;

import static io.github.muehmar.codegenerator.writer.Writer.javaWriter;
import static io.github.muehmar.pojobuilder.generator.impl.gen.Refs.JAVA_LANG_STRING;
import static io.github.muehmar.pojobuilder.generator.impl.gen.Refs.JAVA_UTIL_LIST;
import static io.github.muehmar.pojobuilder.generator.impl.gen.Refs.JAVA_UTIL_MAP;
import static io.github.muehmar.pojobuilder.generator.impl.gen.Refs.JAVA_UTIL_OPTIONAL;
import static io.github.muehmar.pojobuilder.snapshottesting.SnapshotUtil.writerSnapshot;
import static org.assertj.core.api.Assertions.assertThat;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import ch.bluecare.commons.data.PList;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.writer.Writer;
import io.github.muehmar.pojobuilder.generator.PojoFields;
import io.github.muehmar.pojobuilder.generator.Pojos;
import io.github.muehmar.pojobuilder.generator.impl.gen.builder.shared.SetMethodGenerator.ModifierOption;
import io.github.muehmar.pojobuilder.generator.model.Name;
import io.github.muehmar.pojobuilder.generator.model.Pojo;
import io.github.muehmar.pojobuilder.generator.model.settings.PojoSettings;
import io.github.muehmar.pojobuilder.snapshottesting.SnapshotTest;
import org.junit.jupiter.api.Test;

@SnapshotTest
class SetMethodGeneratorTest {
  private Expect expect;

  @Test
  @SnapshotName("requiredField")
  void setMethodGenerator_when_requiredField_then_correctOutput() {
    final Generator<Pojo, PojoSettings> generator =
        SetMethodGenerator.setMethodGenerator((p, s) -> "Builder", ModifierOption.REQUIRED_PRIVATE);

    final Writer writer =
        generator.generate(
            Pojos.fromFields(PojoFields.requiredId()),
            PojoSettings.defaultSettings(),
            javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("requiredFieldAndPublicModifier")
  void setMethodGenerator_when_requiredFieldAndPublicModifier_then_correctOutput() {
    final Generator<Pojo, PojoSettings> generator =
        SetMethodGenerator.setMethodGenerator((p, s) -> "Builder", ModifierOption.PUBLIC);

    final Writer writer =
        generator.generate(
            Pojos.fromFields(PojoFields.requiredId()),
            PojoSettings.defaultSettings(),
            javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("optionalField")
  void setMethodGenerator_when_optionalField_then_correctOutput() {
    final Generator<Pojo, PojoSettings> generator =
        SetMethodGenerator.setMethodGenerator((p, s) -> "Builder", ModifierOption.REQUIRED_PRIVATE);

    final Writer writer =
        generator.generate(
            Pojos.fromFields(PojoFields.optionalId()),
            PojoSettings.defaultSettings(),
            javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  void setMethodGenerator_when_generatorUsedWithGenericType_then_correctRefs() {
    final Generator<Pojo, PojoSettings> generator =
        SetMethodGenerator.setMethodGenerator((p, s) -> "Builder", ModifierOption.REQUIRED_PRIVATE);

    final Writer writer =
        generator.generate(
            Pojos.fromFields(PojoFields.requiredMap()),
            PojoSettings.defaultSettings(),
            javaWriter());

    assertThat(writer.getRefs().exists(JAVA_UTIL_LIST::equals)).isTrue();
    assertThat(writer.getRefs().exists(JAVA_LANG_STRING::equals)).isTrue();
    assertThat(writer.getRefs().exists(JAVA_UTIL_MAP::equals)).isTrue();
  }

  @Test
  @SnapshotName("fieldOfGenericClass")
  void setMethodGenerator_when_fieldOfGenericClass_then_correctParameterizableBuilder() {
    final Generator<Pojo, PojoSettings> generator =
        SetMethodGenerator.setMethodGenerator((p, s) -> "Builder", ModifierOption.REQUIRED_PRIVATE);

    final Writer writer =
        generator.generate(
            Pojos.genericSample().withFields(PList.single(PojoFields.optionalId())),
            PojoSettings.defaultSettings(),
            javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("builderSetMethodePrefix")
  void setMethodGenerator_when_builderSetMethodePrefix_then_correctPublicMethodGenerated() {
    final Generator<Pojo, PojoSettings> generator =
        SetMethodGenerator.setMethodGenerator((p, s) -> "Builder", ModifierOption.REQUIRED_PRIVATE);

    final PojoSettings settings =
        PojoSettings.defaultSettings().withBuilderSetMethodPrefixOpt(Name.fromString("set"));

    final Writer writer =
        generator.generate(Pojos.fromFields(PojoFields.optionalId()), settings, javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("optionalFieldWithGenericType")
  void setMethodGenerator_when_optionalFieldWithGenericType_then_correctRefs() {
    final Generator<Pojo, PojoSettings> generator =
        SetMethodGenerator.setMethodGenerator((p, s) -> "Builder", ModifierOption.REQUIRED_PRIVATE);

    final Writer writer =
        generator.generate(
            Pojos.genericSample().withFields(PList.single(PojoFields.optionalMap())),
            PojoSettings.defaultSettings(),
            javaWriter());

    assertThat(writer.getRefs().exists(JAVA_UTIL_OPTIONAL::equals)).isTrue();
    assertThat(writer.getRefs().exists(JAVA_UTIL_MAP::equals)).isTrue();
    assertThat(writer.getRefs().exists(JAVA_LANG_STRING::equals)).isTrue();
    assertThat(writer.getRefs().exists(JAVA_UTIL_LIST::equals)).isTrue();
  }

  @Test
  @SnapshotName("genericClass")
  void setMethodGenerator_when_genericClass_then_correctPublicMethodGenerated() {
    final Generator<Pojo, PojoSettings> generator =
        SetMethodGenerator.setMethodGenerator((p, s) -> "Builder", ModifierOption.REQUIRED_PRIVATE);

    final Writer writer =
        generator.generate(
            Pojos.genericSample().withFields(PList.single(PojoFields.optionalId())),
            PojoSettings.defaultSettings(),
            javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }
}
