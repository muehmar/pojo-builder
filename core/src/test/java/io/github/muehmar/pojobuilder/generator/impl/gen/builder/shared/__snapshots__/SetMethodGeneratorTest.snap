builderSetMethodePrefix=[
java.lang.Integer
java.util.Optional

public Builder setId(Integer id) {
  this.id = id;
  return this;
}

public Builder setId(Optional<Integer> id) {
  this.id = id.orElse(null);
  return this;
}

]


fieldOfGenericClass=[
java.lang.Integer
java.util.Optional

public Builder<T, S> id(Integer id) {
  this.id = id;
  return this;
}

public Builder<T, S> id(Optional<Integer> id) {
  this.id = id.orElse(null);
  return this;
}

]


genericClass=[
java.lang.Integer
java.util.Optional

public Builder<T, S> id(Integer id) {
  this.id = id;
  return this;
}

public Builder<T, S> id(Optional<Integer> id) {
  this.id = id.orElse(null);
  return this;
}

]


optionalField=[
java.lang.Integer
java.util.Optional

public Builder id(Integer id) {
  this.id = id;
  return this;
}

public Builder id(Optional<Integer> id) {
  this.id = id.orElse(null);
  return this;
}

]


requiredField=[
java.lang.Integer

private Builder id(Integer id) {
  this.id = id;
  return this;
}

]


requiredFieldAndPublicModifier=[
java.lang.Integer

public Builder id(Integer id) {
  this.id = id;
  return this;
}

]